src/StorageWriter/ClaimController.ts

```diff
@@ -1,51 +1,102 @@
 import { Claim } from '@po.et/poet-js'
 import { inject, injectable } from 'inversify'
-import { Collection, Db } from 'mongodb'
 import * as Pino from 'pino'
+import { pipeP } from 'ramda'
 
 import { childWithFileName } from 'Helpers/Logging'
-import { Exchange } from 'Messaging/Messages'
-import { Messaging } from 'Messaging/Messaging'
 
+import { Database } from './Database'
+import { uploadClaimError, updateClaimHashError } from './Errors'
 import { IPFS } from './IPFS'
 
+enum LogTypes {
+  'info' = 'info',
+  'trace' = 'trace',
+  'error' = 'error',
+}
+
+interface StoreNextClaimData {
+  claim: Claim
+  ipfsFileHash?: string
+}
+
 @injectable()
 export class ClaimController {
   private readonly logger: Pino.Logger
-  private readonly db: Db
-  private readonly collection: Collection
-  private readonly messaging: Messaging
+  private readonly db: Database
   private readonly ipfs: IPFS
 
-  constructor(
-    @inject('Logger') logger: Pino.Logger,
-    @inject('DB') db: Db,
-    @inject('Messaging') messaging: Messaging,
-    @inject('IPFS') ipfs: IPFS
-  ) {
+  constructor(@inject('Logger') logger: Pino.Logger, @inject('Database') db: Database, @inject('IPFS') ipfs: IPFS) {
     this.logger = childWithFileName(logger, __filename)
     this.db = db
-    this.collection = this.db.collection('storageWriter')
-    this.messaging = messaging
     this.ipfs = ipfs
   }
 
-  async create(claim: Claim): Promise<void> {
-    const logger = this.logger.child({ method: 'create' })
+  private readonly log = (level: LogTypes) => (message: string) => async (value: any) => {
+    const logger = this.logger
+    logger[level]({ value }, message)
+    return value
+  }
 
-    logger.trace({ claim }, 'Storing Claim')
+  // tslint:disable-next-line
+  public readonly addClaim = async (claim: Claim): Promise<void> => {
+    const logger = this.logger.child({ method: 'addClaim' })
 
-    const ipfsFileHash = await this.ipfs.addText(JSON.stringify(claim))
+    logger.trace({ claim }, 'Adding Claim')
 
-    logger.info({ claim, ipfsFileHash }, 'Claim Stored')
+    await this.db.addClaim(claim)
 
-    await this.collection.insertOne({
-      claimId: claim.id,
-      ipfsFileHash,
-    })
-    await this.messaging.publish(Exchange.ClaimIPFSHash, {
-      claimId: claim.id,
-      ipfsFileHash,
-    })
+    logger.info({ claim }, 'Claim Added')
   }
+
+  private readonly findNextClaim = async (): Promise<StoreNextClaimData> => {
+    const claim = await this.db.findNextClaim()
+    return { claim }
+  }
+
+  private readonly uploadClaim = (claim: Claim) => this.ipfs.addText(JSON.stringify(claim))
+
+  private readonly handleStoreClaimError = async (error: Error, claim: Claim) => {
+    await this.db.addError(uploadClaimError(error.message, claim))
+    throw new Error('Failed to store claim')
```

Comment:
I was thinking it would be better to track each attempt error because it might fail for different reasons.

Reply:
Your comment below makes a good point though, this should be a logger error that gets persisted, not added to the database.

Reply:
Awesome! Responded about this on the other thread.

> I was thinking it would be better to track each failure because it might fail for different reasons and it would be good to see each attempt for debugging.

If you feel there could be actual errors happening here, then yes, let's keep the log at error level and adjust if we see too many false negatives.